### PR TITLE
[Issue 40] FInalization of blocks

### PR DESCRIPTION
**Pull request by pecornilleau**
- The index of the last block in the finalized part of the chain is kept
- Finalization can be requested at any point, by anyone, and entails just checking if the finality period of first children of last finalized is finished.
- In removal, the blocks that are recursively deleted are refunded to those who proposed them

TODO / NEW ISSUE : When finalizing, remove all siblings (block that will no longer be candidates for finalization

### CODE REVIEW

File: layer1/chain/src/chain_sc.mligo [hunk 1]
```diff
@@ -3,24 +3,80 @@
...
+    | Finalize -> apply_finalize store
+    | Remove i -> apply_remove (i, store)
 
-(* VIEWS *)
+(* VIEWS *) 
 
 [@view] let get_chain ((),s: unit * chain_storage) : chain_storage = s
-[@view] let get_latest ((),s: unit * chain_storage) : block option = find_latest_existing s
\ No newline at end of file
+[@view] let get_latest ((),s: unit * chain_storage) : block option = find_latest_existing s
```
Comment (rueshyna) [inline layer1/chain/src/chain_sc.mligo:78]: I missed this before.
`chain_storage` includes `big_map`. It's forbidden for views.

```
File "chain/src/chain_sc.mligo", line 25, characters 23-27:

Invalid view.
Type 'record[blocks -> big_map (nat , record[hash -> bytes , level -> nat , parent -> nat , proposer -> address]) , children -> big_map (nat , list (nat)) , max_index -> nat]' is forbidden as output argument.
```

Comment (pecornilleau) [reply layer1/chain/src/chain_sc.mligo:78]: > I missed this before. `chain_storage` includes `big_map`. It's forbidden in input and output type for views.
> 
> ```
> File "chain/src/chain_sc.mligo", line 25, characters 23-27:
> 
> Invalid view.
> Type 'record[blocks -> big_map (nat , record[hash -> bytes , level -> nat , parent -> nat , proposer -> address]) , children -> big_map (nat , list (nat)) , max_index -> nat]' is forbidden as output argument.
> ```

I clearly didn't take the views seriously enough. So there is no way to add views to a contract with `big_map` storage ? I didn't expect that.

Comment (rueshyna) [reply layer1/chain/src/chain_sc.mligo:78]: oh, I misunderstand about ligo.
so if the type of views is `arg * storage -> return`
`arg` and `return` can't have `big_map` but in `storage` is ok.
you can operate `big_map` in views, just can't return `big_map`

File: layer1/chain/src/chain.mligo [hunk 2]
```diff
@@ -79,3 +128,43 @@ let find_latest_existing (chain : chain) : block option =
...
+let rec get_last_elt (type a) (l : a list) : a option =
+    match l with
+    | [] -> (None : a option)
+    | t::[] -> Some t
+    | t::q -> get_last_elt q 
+
+(** [get_finalization_candidate_index (chain)] returns the index of a block, which is the first registered child of last finalized block *)
+let get_finalization_candidate_index (chain : chain) : index option =
+    match get_children (chain.latest_finalized, chain) with
+        | Some children -> get_last_elt children
```
Comment (rueshyna) [inline layer1/chain/src/chain.mligo:142]: It's ok for now.
but one of assumptions for optimistic rollup is that the head of children is last added and it should be first finalization candidate.

**Review by rueshyna — APPROVED** [review]
leave some minor comments.
LGTM.

File: layer1/chain/src/chain.mligo [hunk 3]
```diff
@@ -2,25 +2,67 @@
...
+    {  index = index
+    ;  parent = proposal.parent
+    ;  level = proposal.level
+    ;  hash = proposal.hash
+    ;  proposer = proposer
+    ;  date_of_proposition = now
+    }
+
+(** [increase_index (chain)] returns a new chain with the max_index increased *)
+let increase_index (chain : chain) : chain =
```
[the remark this answers is not in the document]
Comment (pecornilleau) [reply layer1/chain/src/chain.mligo:62]: Not super satisfied with this part. If you look at `apply_receive` in `chain_sc`, I wanted to have an independent `make_block` function, that takes the proposer and date as argument, but it also need the `index`. So I computed the index before. The sequence is then 
`increase_index`
`make_block`
`store_block`

File: layer1/chain/src/chain.mligo (hunk 2 above)
Comment (pecornilleau) [reply layer1/chain/src/chain.mligo:142]: I added a FIXME to keep that in mind.

**Review by xvw — APPROVED** [review]

**Pull request merged**